clang/lib/CConv/ConstraintResolver.cpp

```diff
@@ -205,6 +205,9 @@ std::set<ConstraintVariable *>
 
     // variable (x)
     } else if (DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(E)) {
+      if(TypE->isStructureType()) {
+        return PVConstraintFromType(TypE);
+      }
```

Comment:
Line 173 checks for "arithmetic type" and returns a base type in that case. Why not just extend that check to include `struct` types as well? Is it because then compound literals will not be handled properly?

Reply:
And if that's the case, why is the approach here actually correct? In other words, why should variables be treated differently than other ways you could return a `struct` type?

Reply:
The reason is that initialization list for structs must be handled explicitly so that constraints for the struct members can be generated. In the `DeclRef`, case the constraints already exist.  Thinking more about this, I'm not too happy about having struct init list expressions making a recursive call to `constrainLocalAssign`. If assignment to struct fields was handled separately, then we could write `    if (TypE->isArithmeticType() || TypeE->isStructureType())` without issue.

Reply:
It looks like my changes for compound literals will touch the struct init list code, so let's hold off on this a bit longer.